ovirt_imageio/tool/options.py

```diff
@@ -0,0 +1,50 @@
+# ovirt-imageio
+# Copyright (C) 2022 Red Hat, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+
+"""
+Tool options.
+"""
+
+import argparse
+
+
+class Parser:
+
+    def __init__(self):
+        self._parser = argparse.ArgumentParser(
+            description="Transfer disk images")
+        self._parser.set_defaults(
+            command=lambda _: self._parser.print_help())
+        self._commands = self._parser.add_subparsers(title="commands")
+
+    def add_sub_command(self, name, help, func):
+        cmd = self._commands.add_parser(name, help=help)
+        cmd.set_defaults(command=func)
+
+        cmd.add_argument(
+            "--engine-url",
+            help="ovirt-engine URL.")
+
+        cmd.add_argument(
+            "--username",
+            help="ovirt-engine username.")
+
+        cmd.add_argument(
+            "--password-file",
+            help="Read ovirt-engine password from file.")
+
+        cmd.add_argument(
+            "--cafile",
+            help="Path to ovirt-engine CA certificate")
```

Comment:
This is a really nice argparse pattern you made here.

But, are these arguments added for each sub command? Can't we have them as global options already in the constructor, so we don't have to add them every time?

Reply:
If we put the arguments on the parser they will not be reported as command options.

See how current backup_vm.py works:

```
$ ./backup_vm.py -h
usage: backup_vm.py [-h] -c CONFIG [--debug] [--logfile LOGFILE] {full,incremental,start,download,stop} ...

Backup VM disks

options:
  -h, --help            show this help message and exit
  -c CONFIG, --config CONFIG
                        Use engine connection details from [CONFIG] section in ~/.config/ovirt.conf.
  --debug               Log debug level messages to logfile.
  --logfile LOGFILE     Log file name (default example.log).

commands:
  {full,incremental,start,download,stop}
    full                Run full backup.
    incremental         Run incremental backup.
    start               Start VM backup.
    download            Download VM backup disk.
    stop                Stop VM backup.
```

```
$ ./backup_vm.py full -h
usage: backup_vm.py full [-h] [--max-workers MAX_WORKERS] [--buffer-size BUFFER_SIZE] [--backup-dir BACKUP_DIR] [--timeout-policy {legacy,pause,cancel}] [--backup-uuid BACKUP_UUID]
                         [--disk-uuid DISK_UUID] [--description DESCRIPTION] [--skip-download]
                         vm_uuid

positional arguments:
  vm_uuid               UUID of the VM to backup.

options:
  -h, --help            show this help message and exit
  --max-workers MAX_WORKERS
                        maximum number of workers to use for backup. The default (4) improves performance when backing up a single disk. You may want to use lower number if you back up
                        many disks in the same time.
  --buffer-size BUFFER_SIZE
                        Buffer size per worker. The default (4194304) gives good performance with the default number of workers. If you use smaller number of workers you may want use
                        larger value.
  --backup-dir BACKUP_DIR
                        Path to a directory to download backup disks to (The default is the current directory).
  --timeout-policy {legacy,pause,cancel}
                        The action to be made for a timed out transfer.
  --backup-uuid BACKUP_UUID
                        UUID of the created VM backup.
  --disk-uuid DISK_UUID
                        Disk UUID to backup. May be used multiple times to backup multiple disks. If not specified, backup all VM disks.
  --description DESCRIPTION
                        A description for the created backup/checkpoint to persist in the Engine DB.
  --skip-download       If specified, start and stop a backup without downloading the disks.
```

You cannot mix gloabl options in the command:

```
$ ./backup_vm.py full -c engine
usage: backup_vm.py [-h] -c CONFIG [--debug] [--logfile LOGFILE] {full,incremental,start,download,stop} ...
backup_vm.py: error: the following arguments are required: -c/--config
```

And you cannot see which options are available for single command.

When we add *all* the options only the sub parser it is much easier to use.

Reply:
Oh snap. Ok thanks for showing.

[Click](https://click.palletsprojects.com/en/8.1.x/) handles this a bit more nicely. Just as a footnote, I know we don’t want to add this dependency, but is a good alternative that I used in the past.